walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -18,30 +18,61 @@
 import java.util.List;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.HashSet;
+import java.util.Stack;
 
 public final class FindMeetingQuery {
 
+  /**
+   * Return list of possible meeting times optimized to include as many optional attendees as possible in 
+   * addition to mandatory attendees.  
+   *
+   * @param events list of events to consider for the attendees requested for the meeting in @param request
+   * @return optimized list of meeting times where ranges accommodate the maximum number of optional attendees
+   */
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long duration = request.getDuration();
+
+    List<TimeRange> unavailableTimes = getUnavailableTimes(events, attendees);
+    List<TimeRange> unavailableTimesOptionalAttendees = getUnavailableTimes(events, optionalAttendees);
+    List<TimeRange> meetingTimes = getMeetingTimes(unavailableTimes, duration);
+    
+    return getMaximallyAccommodatedMeetingTimes(meetingTimes, unavailableTimesOptionalAttendees, duration);
+  }
+
   /** 
    * Return list of unavialable times to schedule the meeting. 
-   * List of TimeRanges is in sorted order from earliest to latest start time.
+   * List of TimeRanges is in sorted order based on the comparator. 
+   *
+   * @param events collection of events to consider for @param attendees unavailable times
+   * @param comparator the comparator function used to sort @return list of unavailable times
    */
   private List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> attendees) {
-    HashSet<TimeRange> unavailableTimesSet = new HashSet<>();
+    List<TimeRange> unavailableTimesList = new ArrayList<>();
     for (Event e : events) {
       Collection<String> eventAttendees = e.getAttendees();
       if (!Collections.disjoint(eventAttendees, attendees)) {
-        unavailableTimesSet.add(e.getWhen());
+        HashSet<String> intersection = new HashSet<>(eventAttendees);
+        intersection.retainAll(attendees);
+        for (int i = 0; i < intersection.size(); i++) {
+          unavailableTimesList.add(e.getWhen());
+        }
```

Comment:
hmm, something's off here
the size of the intersection shouldn't matter, correct?
if all you're doing is adding the same `e.getWhen()` multiple times, no need for the for loop

Reply:
My thinking was to add the timeRange the number of times that there are attendees that we must consider in that event.

Reply:
ah, ok